packages/iov-keycontrol/src/userprofile.spec.ts

```diff
@@ -37,6 +42,275 @@ describe("UserProfile", () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      const createdAt = new ReadonlyDate("1985-04-12T23:20:50.521Z");
+      const keyring = new Keyring();
+      const original = new UserProfile({ createdAt: createdAt, keyring: keyring });
+
+      await original.storeIn(db, defaultEncryptionPassword);
+
+      const otherEncryptionPassword = "something wrong";
+      await UserProfile.loadFrom(db, otherEncryptionPassword)
+        .then(() => fail("loading must not succeed"))
+        .catch(error => expect(error).toMatch(/invalid usage/));
+
+      await db.close();
+    });
+
+    it("throws when loading a profile with no format version", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      await UserProfile.loadFrom(db, defaultEncryptionPassword)
+        .then(() => fail("must not resolve"))
+        .catch(error => expect(error).toMatch(/key not found in database/i));
+    });
+
+    it("throws when loading a profile with unsupported format version", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+      await db.put("format_version", "123");
+
+      await UserProfile.loadFrom(db, defaultEncryptionPassword)
+        .then(() => fail("must not resolve"))
+        .catch(error => expect(error).toMatch(/unsupported format version/i));
+    });
+
+    it("works for password and encryption key", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      const createdAt = new ReadonlyDate("1985-04-12T23:20:50.521Z");
+      const keyring = new Keyring();
+      const original = new UserProfile({ createdAt: createdAt, keyring: keyring });
+      const { id } = original.addWallet(Ed25519HdWallet.fromMnemonic(defaultMnemonic1));
+
+      await original.storeIn(db, defaultEncryptionPassword);
+
+      const encryptionKey = await UserProfile.deriveEncryptionKey(defaultEncryptionPassword);
+      const fromEncryptionKey = await UserProfile.loadFrom(db, encryptionKey);
+      const fromPassword = await UserProfile.loadFrom(db, defaultEncryptionPassword);
+
+      // simulate equality check
+      expect(fromEncryptionKey.createdAt).toEqual(fromPassword.createdAt);
+      expect(fromEncryptionKey.getAllIdentities()).toEqual(fromPassword.getAllIdentities());
+      expect(fromEncryptionKey.printableSecret(id)).toEqual(fromPassword.printableSecret(id));
+
+      await db.close();
+    });
+
+    it("can load format version 1 profile with password", async () => {
+      const db = levelup(MemDownConstructor<string, string>());
+
+      // Storage data created with IOV-Core 0.14 cli
+      await db.put("format_version", userprofileData.serializations.version1.format_version);
+      await db.put("created_at", userprofileData.serializations.version1.created_at);
+      await db.put("keyring", userprofileData.serializations.version1.keyring);
+
+      const loaded = await UserProfile.loadFrom(db, "secret passwd");
```

Comment:
Would be nice to have this stored in a more obvious place. Perhaps even in the JSON file.